### ISSUE TITLE
compiler error notes: Clang constexpr variable lifetime

### ISSUE THREAD

**Issue by alifahrri** [alifahrri/nmtools#198]
clang error: note: read of variable whose lifetime has ended

branch: https://github.com/alifahrri/nmtools/tree/clang-compile-error-constexpr-variable-lifetime

CI log (clang-14)
```
2022-10-01T13:26:22.6894603Z In file included from /work/tests/meta/array/view/moveaxis.cpp:5:
2022-10-01T13:26:22.6895063Z In file included from /work/include/nmtools/array/view/moveaxis.hpp:4:
2022-10-01T13:26:22.6895649Z In file included from /work/include/nmtools/array/index/argsort.hpp:5:
2022-10-01T13:26:22.6919427Z /work/include/nmtools/array/shape.hpp:242:47: error: non-type template argument is not a constant expression
2022-10-01T13:26:22.6939473Z                     }, nmtools_tuple<meta::ct<(size_t)at(fixed_shape,0)>>{});
2022-10-01T13:26:22.6943928Z                                               ^~~~~~~~~~~~~~~~~~~~~~~~~
2022-10-01T13:26:22.6951885Z /work/include/nmtools/array/view/transpose.hpp:42:50: note: in instantiation of function template specialization 'nmtools::shape<true, nmtools::array::ndarray_t<std::array<float, 12>, std::tuple<nmtools::meta::integral_constant<unsigned long, 1>, nmtools::meta::integral_constant<unsigned long, 2>, nmtools::meta::integral_constant<unsigned long, 6>>>>' requested here
2022-10-01T13:26:22.6956189Z         using src_shape_type = decltype(nmtools::shape</*force_constant_index*/true>(meta::declval<array_t>()));
2022-10-01T13:26:22.6959811Z                                                  ^
2022-10-01T13:26:22.6966252Z /work/include/nmtools/array/view/decorator.hpp:129:26: note: in instantiation of template class 'nmtools::view::transpose_t<nmtools::array::ndarray_t<std::array<float, 12>, std::tuple<nmtools::meta::integral_constant<unsigned long, 1>, nmtools::meta::integral_constant<unsigned long, 2>, nmtools::meta::integral_constant<unsigned long, 6>>>, std::tuple<nmtools::meta::integral_constant<unsigned long, 1>, nmtools::meta::integral_constant<unsigned long, 0>, nmtools::meta::integral_constant<unsigned long, 2>>>' requested here
2022-10-01T13:26:22.6971507Z     struct decorator_t : view_t<Ts...>
2022-10-01T13:26:22.6971763Z                          ^
2022-10-01T13:26:22.6983224Z /work/include/nmtools/array/view/transpose.hpp:100:16: note: in instantiation of template class 'nmtools::view::decorator_t<nmtools::view::transpose_t, nmtools::array::ndarray_t<std::array<float, 12>, std::tuple<nmtools::meta::integral_constant<unsigned long, 1>, nmtools::meta::integral_constant<unsigned long, 2>, nmtools::meta::integral_constant<unsigned long, 6>>>, std::tuple<nmtools::meta::integral_constant<unsigned long, 1>, nmtools::meta::integral_constant<unsigned long, 0>, nmtools::meta::integral_constant<unsigned long, 2>>>' requested here
2022-10-01T13:26:22.6984659Z         return decorator_t<transpose_t,array_t,axes_t>{{array,axes}};
2022-10-01T13:26:22.6984952Z                ^
2022-10-01T13:26:22.6992103Z /work/include/nmtools/array/view/moveaxis.hpp:37:56: note: in instantiation of function template specialization 'nmtools::view::transpose<nmtools::array::ndarray_t<std::array<float, 12>, std::tuple<nmtools::meta::integral_constant<unsigned long, 1>, nmtools::meta::integral_constant<unsigned long, 2>, nmtools::meta::integral_constant<unsigned long, 6>>>, std::tuple<nmtools::meta::integral_constant<unsigned long, 1>, nmtools::meta::integral_constant<unsigned long, 0>, nmtools::meta::integral_constant<unsigned long, 2>>>' requested here
2022-10-01T13:26:22.6994480Z         using result_t = meta::remove_cvref_t<decltype(transpose(array,*order))>;
2022-10-01T13:26:22.6995209Z                                                        ^
2022-10-01T13:26:22.7005677Z /work/tests/meta/array/view/moveaxis.cpp:54:44: note: in instantiation of function template specialization 'nmtools::view::moveaxis<nmtools::array::ndarray_t<std::array<float, 12>, std::tuple<nmtools::meta::integral_constant<unsigned long, 1>, nmtools::meta::integral_constant<unsigned long, 2>, nmtools::meta::integral_constant<unsigned long, 6>>>, nmtools::meta::integral_constant<unsigned long, 0>, nmtools::meta::integral_constant<unsigned long, 1>>' requested here
2022-10-01T13:26:22.7006825Z         using view_type   = decltype(view::moveaxis(meta::declval<array_type>(),meta::declval<src_type>(),meta::declval<dst_type>()));
2022-10-01T13:26:22.7008561Z                                            ^
2022-10-01T13:26:22.7013855Z /work/include/nmtools/array/shape.hpp:242:55: note: read of variable whose lifetime has ended
2022-10-01T13:26:22.7015592Z                     }, nmtools_tuple<meta::ct<(size_t)at(fixed_shape,0)>>{});
2022-10-01T13:26:22.7017908Z                                                       ^
```

works fine on gcc

related code: https://github.com/alifahrri/nmtools/blob/clang-compile-error-constexpr-variable-lifetime/include/nmtools/array/shape.hpp#L236
```
                constexpr auto fixed_shape = meta::fixed_shape_v<array_t>;
                if constexpr (!meta::is_fail_v<decltype(fixed_shape)>) {
                    return meta::template_reduce<len(fixed_shape)-1>([&](auto init, auto index){
                        using init_type = decltype(init);
                        using type = meta::append_type_t<init_type,meta::ct<(size_t)at(fixed_shape,index+1)>>;
                        return type{};
                    }, nmtools_tuple<meta::ct<(size_t)at(fixed_shape,0)>>{});
                } else {
                    return impl::shape<array_t>(array);
                }
```

**Comment by alifahrri** [comment]
error message: https://github.com/alifahrri/nmtools/actions/runs/3811800301/jobs/6484666756 
```
/work/tests/array/index/reshape.cpp:639:5: error: constexpr variable 'result' must be initialized by a constant expression
    CONSTEXPR_SHAPE_RESHAPE_SUBCASE( case5, shape, newshape_ct );
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/work/tests/array/index/reshape.cpp:377:20: note: expanded from macro 'CONSTEXPR_SHAPE_RESHAPE_SUBCASE'
    constexpr auto result = RUN_impl(__VA_ARGS__); \
                   ^        ~~~~~~~~~~~~~~~~~~~~~
/work/include/nmtools/array/index/reshape.hpp:21:30: note: read of variable whose lifetime has ended
            const auto d_i = at(dst_shape,i);
                             ^
/work/include/nmtools/array/index/reshape.hpp:77:53: note: in call to 'count_negative_reshape(newshape_ct)'
            const auto [minus_1_count, dst_numel] = count_negative_reshape(dst_shape);
                                                    ^
```

related code: https://github.com/alifahrri/nmtools/blob/clang-compile-error-constexpr-variable-lifetime-2/include/nmtools/array/index/reshape.hpp#L21 

```
    template <typename dst_shape_t>
    constexpr auto count_negative_reshape(const dst_shape_t& dst_shape)
    {
        auto minus_1_count = 0;
        auto dst_numel = (size_t)0;
        using index_t = meta::get_index_element_type_t<dst_shape_t>;
        for (size_t i=0; i<len(dst_shape); i++) {
            const auto d_i = at(dst_shape,i);
            if (i==0) {
                dst_numel = 1;
            }
```

**Comment by alifahrri** [comment]
fixed in #216
